drivers/rtc/rtc_ds1307.c

```diff
@@ -0,0 +1,167 @@
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	LOG_DBG("set time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		tm->tm_year, tm->tm_mon, tm->tm_mday, tm->tm_wday, tm->tm_hour, tm->tm_min,
+		tm->tm_sec);
+
+	regs[0] = bin2bcd(tm->tm_sec) & SECONDS_BITS;
+	regs[1] = bin2bcd(tm->tm_min);
+	regs[2] = bin2bcd(tm->tm_hour);
+	regs[3] = bin2bcd(tm->tm_wday);
+	regs[4] = bin2bcd(tm->tm_mday);
+	regs[5] = bin2bcd(tm->tm_mon);
+	regs[6] = bin2bcd((tm->tm_year % 100));
+
+	err = i2c_burst_write_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+
+	k_spin_unlock(&data->lock, key);
+
+	return err;
+}
+
+static int ds1307_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	int err;
+	uint8_t regs[7];
+
+	struct ds1307_data *data = dev->data;
+	const struct ds1307_config *config = dev->config;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	err = i2c_burst_read_dt(&config->i2c_bus, DS1307_REG_SECONDS, regs, sizeof(regs));
+	if (err != 0) {
+		goto unlock;
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+	timeptr->tm_year = timeptr->tm_year + 100;
+
+	/* Not used */
+	timeptr->tm_nsec =  0;
+	timeptr->tm_isdst =  -1;
+	timeptr->tm_yday = -1;
+
+	/* Validate the chip in 24hr mode */
+	if (regs[2] & (1 << 6)) {
+		return -ENODATA;
+	}
+
+	LOG_DBG("get time: year = %d, mon = %d, mday = %d, wday = %d, hour = %d, "
+		"min = %d, sec = %d",
+		timeptr->tm_year, timeptr->tm_mon, timeptr->tm_mday, timeptr->tm_wday,
+		timeptr->tm_hour, timeptr->tm_min, timeptr->tm_sec);
+
+unlock:
+	k_spin_unlock(&data->lock, key);
+
+	return err;
+}
+
+static const struct rtc_driver_api ds1307_driver_api = {
+	.set_time = ds1307_set_time,
+	.get_time = ds1307_get_time,
+};
+
+static int ds1307_init(const struct device *dev)
+{
+	int err;
+	uint8_t tx_buf[2];
+	const struct ds1307_config *config = dev->config;
+
+	if (!i2c_is_ready_dt(&config->i2c_bus)) {
+		LOG_ERR("I2C bus not ready");
+		return -ENODEV;
+	}
+
+	tx_buf[0] = DS1307_REG_CTRL;
+	tx_buf[1] = 0x00;
+
+	/* Disable squarewave output */
+	err = i2c_write_dt(&config->i2c_bus, tx_buf, sizeof(tx_buf));
+	if (err < 0) {
+		LOG_ERR("Error: SQW:%d\n", err);
+	}
+
+	tx_buf[0] = DS1307_REG_SECONDS;
+
+	/* Make clock halt = 0 */
+	err = i2c_write_dt(&config->i2c_bus, tx_buf, sizeof(tx_buf));
+	if (err < 0) {
+		LOG_ERR("Error: Set clock halt bit:%d\n", err);
+	}
+
+	return 0;
+}
+
+#define DS1307_INIT(inst)                                                                          \
```

Comment:
nit: these are normally named something like `DS1307_DEFINE`, init is the initialization function, this is defining a bunch of stuff.